hpke/xkem.go

```diff
@@ -0,0 +1,144 @@
+package hpke
+
+import (
+	"bytes"
+	"crypto/rand"
+	_ "crypto/sha256" // linking sha256 packages.
+	_ "crypto/sha512" // linking sha512 packages.
+	"crypto/subtle"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/cloudflare/circl/dh/x25519"
+	"github.com/cloudflare/circl/dh/x448"
+	"github.com/cloudflare/circl/kem"
+)
+
+type xkem struct {
+	kemBase
+	size int
+}
+
+func (x xkem) PrivateKeySize() int        { return x.size }
+func (x xkem) SeedSize() int              { return x.size }
+func (x xkem) CiphertextSize() int        { return x.size }
+func (x xkem) PublicKeySize() int         { return x.size }
+func (x xkem) EncapsulationSeedSize() int { return x.size }
+
+func (x xkem) sizeDH() int { return x.size }
+func (x xkem) calcDH(dh []byte, sk kem.PrivateKey, pk kem.PublicKey) error {
+	PK := pk.(*xkemPubKey)
+	SK := sk.(*xkemPrivKey)
+	switch x.size {
+	case x25519.Size:
+		var ss, sKey, pKey x25519.Key
+		copy(sKey[:], SK.k)
+		copy(pKey[:], PK.k)
+		x25519.Shared(&ss, &sKey, &pKey)
+		copy(dh, ss[:])
+	case x448.Size:
+		var ss, sKey, pKey x448.Key
+		copy(sKey[:], SK.k)
+		copy(pKey[:], PK.k)
+		x448.Shared(&ss, &sKey, &pKey)
+		copy(dh, ss[:])
+	}
+	return nil
+}
+func (x xkem) DeriveKey(seed []byte) (kem.PublicKey, kem.PrivateKey) {
+	// Implementation based on
+	// https://www.ietf.org/archive/id/draft-irtf-cfrg-hpke-06.html#name-derivekeypair
+	if len(seed) != x.SeedSize() {
+		panic(kem.ErrSeedSize)
```

Comment:
Eh, on second thought I think this is fine. However, we need to document this behavior.

Reply:
yes, it satisy the `kem.Scheme` interface, which announces this panic.

Reply:
It's also worth mentioning here. Imagine you were looking at the go docs for this function -- I don't think it would be immediately apparent that this object implements `kem.Scheme`.